src/main/java/com/project/store/controller/product/ProductController.java

```diff
@@ -0,0 +1,70 @@
+package com.project.store.controller.product;
+
+import com.project.store.dto.category.CategoryDto;
+import com.project.store.dto.product.ProductDto;
+import com.project.store.mapper.ProductMapper;
+import com.project.store.model.category.Category;
+import com.project.store.model.product.Product;
+import com.project.store.service.impl.CategoryServiceImpl;
+import com.project.store.service.impl.ProductServiceImpl;
+import lombok.RequiredArgsConstructor;
+import org.springframework.web.bind.annotation.*;
+
+import javax.validation.Valid;
+import java.util.List;
+
+@RestController
+@RequestMapping(value = "/product")
+@RequiredArgsConstructor
+public class ProductController {
+
+    private final ProductServiceImpl productService;
+    private final ProductMapper productMapper;
+    private final CategoryServiceImpl categoryService;
+
+    @PostMapping()
+    public ProductDto addProduct(@Valid @RequestBody ProductDto productDto) {
+        Category category = categoryService.getCategory(productDto.getCategoryId());
+        Product product = productMapper.fromDtoToEntity(productDto, category);
+        Product productToBeAdded = productService.addProduct(product);
+
+        return productMapper.fromEntityToDto(productToBeAdded);
+    }
+
+    @GetMapping("/{id}")
+    public ProductDto getProduct(@PathVariable Integer id) {
+        Product product = productService.getProduct(id);
+
+        return productMapper.fromEntityToDto(product);
+    }
+
+    @GetMapping()
+    public List<ProductDto> getAllProducts() {
+        List<Product> productList = productService.getAllProducts();
+        return productMapper.fromEntitiesToDtos(productList);
+    }
+
+    @PutMapping()
+    public ProductDto updateProduct(@Valid @RequestBody ProductDto productDto) {
+        Product productToBeUpdated = productMapper.fromDtoToEntity(productDto);
+        Product updatedProduct = productService.updateProduct(productToBeUpdated);
+
+        return productMapper.fromEntityToDto(updatedProduct);
+    }
+
+    @PutMapping("/product-and-category")
+    public ProductDto updateProductAndCategory(@Valid @RequestBody ProductDto productDto) {
+        Category category = categoryService.getCategory(productDto.getCategoryId());
```

Comment:
Can category id be null? Should it be nullable? If so, what happens in this scenario?

Reply:
If an invalid categoryId is sent, then a CategoryNotFoundException is thrown with HttpStatus.404
It's a safety mechanism for not sending a wrong categoryIds

Reply:
In cases like this, it's better to catch the invalid id until we query the database, or do any kind of logic with it. If category id si null, an exception should be thrown from the controller level (eg: using spring validators).
If you propagate a null object (that should not be null) downstream, you risk issues like NullPointerExceptions or other unwanted behaviour.

Reply:
Added @NotNull validation to categoryId from CategoryDto